File: ios/Classes/FlutterBugfenderPlugin.m
#import "FlutterBugfenderPlugin.h"
@import BugfenderSDK;

@implementation FlutterBugfenderPlugin
+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel
            methodChannelWithName:@"flutter_bugfender"
                  binaryMessenger:[registrar messenger]];
    FlutterBugfenderPlugin *instance = [[FlutterBugfenderPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (BFLogLevel)intToLogLevel:(int) logLevelIndex {
    switch (logLevelIndex) {
        case 0:
            return BFLogLevelTrace;
            break;
        case 1:
            return BFLogLevelDefault;
            break;
        case 2:
            return BFLogLevelInfo;
            break;
        case 3:
            return BFLogLevelWarning;
            break;
        case 4:
            return BFLogLevelError;
            break;
        case 5:
            return BFLogLevelFatal;
            break;
        default:
            return BFLogLevelDefault;
            break;
    }
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"init" isEqualToString:call.method]) {
        NSString *appKey = call.arguments[@"appKey"];
        NSString *apiUri = call.arguments[@"apiUri"];
        NSString *baseUri = call.arguments[@"baseUri"];
        long maximumLocalStorageSize = [call.arguments[@"maximumLocalStorageSize"] longValue];
        BOOL printToConsole = [call.arguments[@"printToConsole"] boolValue];
        BOOL enableUIEventLogging = [call.arguments[@"enableUIEventLogging"] boolValue];
        BOOL enableCrashReporting = [call.arguments[@"enableCrashReporting"] boolValue];
        NSString *overrideDeviceName = call.arguments[@"overrideDeviceName"];

        if (overrideDeviceName.length)
            [Bugfender overrideDeviceName:overrideDeviceName];
        if (apiUri.length)
            [Bugfender setApiURL:[NSURL URLWithString:apiUri]];
        if (baseUri.length)
            [Bugfender setBaseURL:[NSURL URLWithString:baseUri]];
        [Bugfender activateLogger:appKey];
        if (enableUIEventLogging)
            [Bugfender enableUIEventLogging];
        if (enableCrashReporting)
            [Bugfender enableCrashReporting];
        [Bugfender setPrintToConsole:printToConsole];
        if (maximumLocalStorageSize) {
            [Bugfender setMaximumLocalStorageSize:maximumLocalStorageSize];
        }
        result(nil);
    } else if ([@"setDeviceString" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *key = arguments[@"key"];
        NSString *value = arguments[@"value"];
        [Bugfender setDeviceString:value forKey:key];
        result(nil);
    } else if ([@"setDeviceInt" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *key = arguments[@"key"];
        NSInteger value = [arguments[@"value"] integerValue];
        [Bugfender setDeviceInteger:value forKey:key];
        result(nil);
    } else if ([@"setDeviceFloat" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *key = arguments[@"key"];
        CGFloat value = [arguments[@"value"] floatValue];
        [Bugfender setDeviceDouble:value forKey:key];
        result(nil);
    } else if ([@"setDeviceBool" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *key = arguments[@"key"];
        BOOL value = [arguments[@"value"] boolValue];
        [Bugfender setDeviceBOOL:value forKey:key];
        result(nil);
    } else if ([@"removeDeviceKey" isEqualToString:call.method]) {
        NSString *key = call.arguments;
        [Bugfender removeDeviceKey:key];
        result(nil);
    } else if ([@"setForceEnabled" isEqualToString:call.method]) {
        BOOL enable = [call.arguments boolValue];
        [Bugfender setForceEnabled:enable];
        result(nil);
    } else if ([@"forceSendOnce" isEqualToString:call.method]) {
        [Bugfender forceSendOnce];
        result(nil);
    } else if ([@"sendCrash" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *title = arguments[@"title"];
        NSString *value = arguments[@"value"];
        NSURL* url = [Bugfender sendCrashWithTitle:title text:value];
        result(url.absoluteString);
    } else if ([@"sendIssue" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *title = arguments[@"title"];
        NSString *value = arguments[@"value"];
        NSURL* url = [Bugfender sendIssueReturningUrlWithTitle:title text:value];
        result(url.absoluteString);
    } else if ([@"sendUserFeedback" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *title = arguments[@"title"];
        NSString *value = arguments[@"value"];
        NSURL* url = [Bugfender sendUserFeedbackReturningUrlWithSubject:title message:value];
        result(url.absoluteString);
    } else if ([@"getDeviceUri" isEqualToString:call.method]) {
        NSURL* url = [Bugfender deviceIdentifierUrl];
        result(url.absoluteString);
    } else if ([@"getSessionUri" isEqualToString:call.method]) {
        NSURL* url = [Bugfender sessionIdentifierUrl];
        result(url.absoluteString);
    } else if ([@"sendLog" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *lineNumber = arguments[@"line"];
        NSString *method = arguments[@"method"];
        NSString *file = arguments[@"file"];
        NSString *levelOrdinal = arguments[@"level"];
        NSString *tag = arguments[@"tag"];
        NSString *text = arguments[@"text"];
        [Bugfender logWithLineNumber: [lineNumber integerValue] method: method file: file level: [self intToLogLevel: [levelOrdinal integerValue]] tag: tag message: text];
        result(nil);
    } else if ([@"log" isEqualToString:call.method]) {
        BFLog (@"%@", call.arguments);
        result(nil);
    } else if ([@"fatal" isEqualToString:call.method]) {
        BFLogFatal(@"%@", call.arguments);
        result(nil);
    } else if ([@"error" isEqualToString:call.method]) {
        BFLogErr(@"%@", call.arguments);
        result(nil);
    } else if ([@"warn" isEqualToString:call.method]) {
        BFLogWarn(@"%@", call.arguments);
        result(nil);
    } else if ([@"info" isEqualToString:call.method]) {
        BFLogInfo(@"%@", call.arguments);
        result(nil);
    } else if ([@"debug" isEqualToString:call.method]) {
        BFLog(@"%@", call.arguments);
        result(nil);
    } else if ([@"trace" isEqualToString:call.method]) {
        BFLogTrace(@"%@", call.arguments);
        result(nil);
    } else if ([@"getUserFeedback" isEqualToString:call.method]) {
        NSDictionary *arguments = call.arguments;
        NSString *title = arguments[@"title"];
        NSString *hint = arguments[@"hint"];
        NSString *subjectHint = arguments[@"subjectHint"];
        NSString *messageHint = arguments[@"messageHint"];
        NSString *sendButtonText = arguments[@"sendButtonText"];
        NSString *cancelButtonText = arguments[@"cancelButtonText"];

        UIViewController * userFeedbackViewController = [Bugfender userFeedbackViewControllerWithTitle:title hint:hint subjectPlaceholder:subjectHint
        messagePlaceholder:messageHint sendButtonTitle:sendButtonText cancelButtonTitle:cancelButtonText completion:^(BOOL feedbackSent, NSURL * _Nullable url) {
            if (feedbackSent) {
                result(url.absoluteString);
            } else {
                result(nil);
            }
        }];
        userFeedbackViewController.modalPresentationStyle = UIModalPresentationFullScreen;
        //if (@available(iOS 13.0, *)) {
        //    userFeedbackViewController.modalInPresentation = YES;
        //}
        UIViewController* rootViewController = [[[[UIApplication sharedApplication]delegate] window] rootViewController];
        [rootViewController presentViewController:userFeedbackViewController animated:YES completion:nil];
    } else {
        result(FlutterMethodNotImplemented);
    }
}
@end

File: android/src/main/java/com/ninja/flutterbugfender/FlutterBugfenderPlugin.java
package com.ninja.flutterbugfender;

import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.bugfender.android.BuildConfig;
import com.bugfender.sdk.Bugfender;

import android.app.Activity;

import java.net.URL;

/**
 * FlutterBugfenderPlugin
 */
public class FlutterBugfenderPlugin implements MethodCallHandler {

    private final Activity activity;

    private FlutterBugfenderPlugin(Activity activity) {
        this.activity = activity;
    }


    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_bugfender");
        channel.setMethodCallHandler(new FlutterBugfenderPlugin(registrar.activity()));
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        switch (call.method) {
            case "init":
                String appKey = call.argument("appKey");
                String apiUri = call.argument("apiUri");
                String baseUri = call.argument("baseUri");
                int maximumLocalStorageSize = call.argument("maximumLocalStorageSize");
                boolean printToConsole = call.argument("printToConsole");
                boolean enableUIEventLogging = call.argument("enableUIEventLogging");
                boolean enableCrashReporting = call.argument("enableCrashReporting");
                boolean enableAndroidLogcatLogging = call.argument("enableAndroidLogcatLogging");
                String overrideDeviceName = call.argument("overrideDeviceName");

                if (overrideDeviceName != "")
                    Bugfender.overrideDeviceName(overrideDeviceName);
                if (apiUri != "")
                    Bugfender.setApiUrl(apiUri);
                if (baseUri != "")
                    Bugfender.setBaseUrl(baseUri);
                Bugfender.init(activity.getApplicationContext(), appKey, printToConsole);
                if (enableAndroidLogcatLogging)
                    Bugfender.enableLogcatLogging();
                if (enableUIEventLogging)
                    Bugfender.enableUIEventLogging(activity.getApplication());
                if (enableCrashReporting)
                    Bugfender.enableCrashReporting();
                if (maximumLocalStorageSize != 0) {
                    Bugfender.setMaximumLocalStorageSize(maximumLocalStorageSize);
                }
                result.success(null);
                break;
            case "setDeviceString":
            case "setDeviceInt":
            case "setDeviceFloat":
            case "setDeviceBool":
                String key = call.argument("key");
                switch (call.method) {
                    case "setDeviceString":
                        String strvalue = call.argument("value");
                        Bugfender.setDeviceString(key, strvalue);
                        break;
                    case "setDeviceInt":
                        int intvalue = call.argument("value");
                        Bugfender.setDeviceInteger(key, intvalue);
                        break;
                    case "setDeviceFloat":
                        double floatvalue = call.argument("value");
                        Bugfender.setDeviceFloat(key, (float)floatvalue); // losing precision here: flutter's native type is a double but Bugfender only supports float
                        break;
                    case "setDeviceBool":
                        boolean boolvalue = call.argument("value");
                        Bugfender.setDeviceBoolean(key, boolvalue);
                        break;
                }
                result.success(null);
                break;
            case "removeDeviceKey":
                String key_to_remove = call.arguments();
                Bugfender.removeDeviceKey(key_to_remove);
                result.success(null);
                break;
            case "sendCrash":
            case "sendIssue":
            case "sendUserFeedback":
                String title = call.argument("title");
                String issue_val = call.argument("value");
                URL url = null;
                switch (call.method) {
                    case "sendCrash":
                        url = Bugfender.sendCrash(title, issue_val);
                        break;
                    case "sendIssue":
                        url = Bugfender.sendIssue(title, issue_val);
                        break;
                    case "sendUserFeedback":
                        url = Bugfender.sendUserFeedback(title, issue_val);
                        break;
                }
                result.success(url.toString());
                break;
            case "setForceEnabled":
                Boolean enabled = call.arguments();
                Bugfender.setForceEnabled(enabled);
                result.success(null);
                break;
            case "forceSendOnce":
                Bugfender.forceSendOnce();
                result.success(null);
                break;
            case "getDeviceUri":
                result.success(Bugfender.getDeviceUrl().toString());
                break;
            case "getSessionUri":
                result.success(Bugfender.getSessionUrl().toString());
                break;
            case "log":
            case "fatal":
            case "error":
            case "warn":
            case "info":
            case "debug":
            case "trace":
                String log = call.arguments();
                switch(call.method) {
                    case "fatal":
                        Bugfender.f("", log);
                        break;
                    case "error":
                        Bugfender.e("", log);
                        break;
                    case "warn":
                        Bugfender.w("", log);
                        break;
                    case "info":
                        Bugfender.i("", log);
                        break;
                    case "trace":
                        Bugfender.t("", log);
                        break;
                    default:
                        Bugfender.d("", log);
                        break;
                }
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }
}
